gnome-image-installer/pages/confirm/gis-confirm-page.c

```diff
@@ -0,0 +1,281 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
+ * 02111-1307, USA.
+ *
+ */
+
+/* Confirm page {{{1 */
+
+#define PAGE_ID "confirm"
+
+#include "config.h"
+#include "confirm-resources.h"
+#include "gis-confirm-page.h"
+#include "gis-store.h"
+
+#include <udisks/udisks.h>
+#include <glib/gstdio.h>
+#include <glib/gi18n.h>
+#include <gio/gio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+struct _GisConfirmPagePrivate {
+    GtkLabel *vendor_label;
+    GtkLabel *product_label;
+
+    GtkLabel *image1_label;
+    GtkLabel *model1_label;
+    GtkLabel *device1_label;
+    GtkLabel *size1_label;
+
+    GtkBox *warning_box;
+    GtkLabel *warning_label;
+
+    GtkBox *wrong_computer_box;
+
+    GtkLabel *countdown_label;
+    GtkButton *cancel_button;
+    GtkButton *continue_button;
+
+    guint countdown_source;
+    guint countdown_remaining;
+};
+typedef struct _GisConfirmPagePrivate GisConfirmPagePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisConfirmPage, gis_confirm_page, GIS_TYPE_PAGE);
+G_DEFINE_QUARK(confirm-error, gis_confirm_error);
+
+static void
+gis_confirm_page_advance (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GisPage *page = GIS_PAGE (self);
+
+  if (priv->countdown_source != 0)
+    {
+      g_source_remove (priv->countdown_source);
+      priv->countdown_source = 0;
+    }
+
+  gis_assistant_next_page (gis_driver_get_assistant (page->driver));
+}
+
+static gboolean
+gis_confirm_page_countdown_cb (gpointer data)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (data);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  if (priv->countdown_remaining == 0)
+    {
+      gis_confirm_page_advance (self);
+      return G_SOURCE_REMOVE;
+    }
+  else
+    {
+      g_autofree gchar *countdown_message =
+        g_strdup_printf (_("Installation will commence automatically in %d seconds."),
+                         priv->countdown_remaining);
+      gtk_label_set_text (priv->countdown_label, countdown_message);
+      priv->countdown_remaining--;
+      return G_SOURCE_CONTINUE;
+    }
+}
+
+static void
+gis_confirm_page_start_countdown (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  priv->countdown_remaining = 30;
+  gis_confirm_page_countdown_cb (self);
+
+  priv->countdown_source = g_timeout_add_seconds (1,
+                                                  gis_confirm_page_countdown_cb,
+                                                  self);
+}
+
+static void
+gis_confirm_page_shown (GisPage *page)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (page);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GFile *image_file;
+  g_autofree gchar *image_basename = NULL;
+  UDisksClient *udisks_client;
+  UDisksBlock *target_block;
+  UDisksDrive *target_drive = NULL;
```

Comment:
Should `target_drive` be `g_autoptr()`? (`udisks_client_get_drive_for_block()` is transfer full.)

Reply:
It should! But udisks (at least the version in our distro) does not support these macros. Workarounds would include using a variable of some weaker type, or defining the glue inline, but I think it's clearer to just `g_clear_object()` at the earliest possible point (which we do).

Reply:
Perhaps leave it as it is in this branch, but whack a `FIXME` on there and make sure there’s an upstream bug filed in udisks about adding `g_autoptr()` support (if not done already).

Reply:
Interesting – `UDisksDrive` comes from the output of `gdbus-codegen`, which can be configured to generate G_AUTO goop, but it does not appear to emit these definitions for `UDisksDrive`, only for:

```

Reply:
More investigation required, if have a bit of time to take a look? I am currently down a GLib rabbit hole for ramcq.

Reply:
Is that with `gdbus-codegen`’s `--c-generate-autocleanup` argument, or `--c-generate-autocleanup=all`? The default is `--c-generate-autocleanup=objects`.

See https://bugzilla.gnome.org/show_bug.cgi?id=763379 for the history behind this option.

Reply:
Thanks for the pointer. udisks2 is using the default (ie, it does not pass `--c-generate-autocleanup` at all), and depends on too old a release of GLib (2.36) for this argument to be available. Joy! I'll open a PR there for discussion.

Reply:
https://github.com/storaged-project/udisks/pull/492